File: Problems_and_Solutions/days-10 to 19/day-13/Java/Solution4.java
import java.util.Stack;

class Solution4 {
    public static boolean isValid1(String s) {
        while (true) {
            if (s.contains("()")) {
                s = s.replace("()", "");
            } else if (s.contains("{}")) {
                s = s.replace("{}", "");
            } else if (s.contains("[]")) {
                s = s.replace("[]", "");
            } else {
                return s.isEmpty();
            }
        }
    }

    public static boolean isValid2(String s) {
        Stack<Character> st = new Stack<>();
        for (char c : s.toCharArray()) {
            if (c == '(' || c == '[' || c == '{'){
                st.push(c);
            } else {
                if (st.isEmpty()) return false;
                char top = st.pop();
                if (c == ')' && top != '(') {
                    return false;
                }
                if (c == ']' && top != '[') {
                    return false;
                }
                if (c == '}' && top != '{') {
                    return false;
                }
            }
        }
        return st.isEmpty();
    }

    public static void main(String[] args) {
        String s = "()[]{}";
        System.out.println(isValid1(s));
    }
}

File: Problems_and_Solutions/days-10 to 19/day-13/c++/solution 4.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> chars=stack<char>{};
        chars.push('n');
        char topEle='a'; 
        chars.push(s[0]);
        for(int i=1;i<s.size();i++)
        {
            topEle=chars.top();
            if(s[i]=='(' || s[i]=='[' || s[i]=='{')
                chars.push(s[i]);
            else{
                if(s[i]==')' && topEle=='(')
                    chars.pop();
                else if(s[i]=='}' && topEle=='{')
                    chars.pop();
                else if(s[i]==']' && topEle=='[')
                    chars.pop();
                else
                    return false;
            }
        }
        return chars.size()==1;
    }
};